tsfc/loopy.py

```diff
@@ -40,27 +43,70 @@ def pym_multiindex(self, multiindex):
                 indices.append(index)
         return tuple(indices)
 
+    # Generate index from gem multiindex
+    def gem_to_pym_multiindex(self, multiindex):
+        indices = []
+        for index in multiindex:
+            assert index.extent
+            if not index.name:
+                name = self.name_gen(self.index_names[index])
+            else:
+                name = index.name
+            self.index_extent[name] = index.extent
+            indices.append(p.Variable(name))
+        return tuple(indices)
+
+    # Generate index from shape
+    def pymbolic_multiindex(self, shape):
+        indices = []
+        for extent in shape:
+            name = self.name_gen(self.index_names[extent])
+            self.index_extent[name] = extent
+            indices.append(p.Variable(name))
+        return tuple(indices)
+
+    # Generate pym variable from gem
+    def pymbolic_variable_and_destruct(self, node):
+        pym = self.pymbolic_variable(node)
+        if isinstance(pym, p.Subscript):
+            return pym.aggregate, pym.index_tuple
+        else:
+            return pym, ()
+
+    # Generate pym variable or subscript
     def pymbolic_variable(self, node):
+        pym = self._gem_to_pym_var(node)
+        if node in self.indices:
+            indices = self.fetch_multiindex(self.indices[node])
+            if indices:
+                return p.Subscript(pym, indices)
+        return pym
+
+    def _gem_to_pym_var(self, node):
         try:
             pym = self.gem_to_pymbolic[node]
         except KeyError:
             name = self.name_gen(node.name)
             pym = p.Variable(name)
             self.gem_to_pymbolic[node] = pym
-        if node in self.indices:
-            indices = self.pym_multiindex(self.indices[node])
-            if indices:
-                return p.Subscript(pym, indices)
-            else:
-                return pym
-        else:
-            return pym
+        return pym
 
     def active_inames(self):
         # Return all active indices
         return frozenset([i.name for i in self.active_indices.values()])
 
 
+@contextmanager
+def active_indices(mapping, ctx):
+    """Push active indices onto context.
+   :arg mapping: dict mapping gem indices to pymbolic index expressions
+   :arg ctx: code generation context.
+   :returns: new code generation context."""
+    ctx.active_indices.update(mapping)
+    yield ctx
+    [ctx.active_indices.pop(key) for key in mapping.keys()]
```

Comment:
Sorry, I didn't explain well. I wasn't worried about the lines of code. I'm happy with popping, but can you do this as:
```python
for key in mapping:
    ctx.active_indices.pop(key)
```
Please?

Otherwise some later reader of the code might be confused why this side-effectful list construction is not returned anywhere.

Reply:
Yes sure I agree, I thought you would prefer a one-liner because you had removed my loops in your earlier suggestion.